pay/contracts/subscription/HectorSubscription.sol

```diff
@@ -0,0 +1,319 @@
+            uint256 balance = balanceOf[from][plan.token];
+
+            if (balance >= amount) {
+                isActiveForNow = true;
+                chargeAmount = 0;
+            } else {
+                isActiveForNow = false;
+                chargeAmount = amount - balance;
+            }
+        }
+    }
+
+    /* ======== USER FUNCTIONS ======== */
+
+    function deposit(address _token, uint256 _amount) public nonReentrant {
+        IERC20(_token).safeTransferFrom(msg.sender, address(this), _amount);
+
+        emit PayerDeposit(msg.sender, _token, _amount);
+    }
+
+    function createSubscription(uint256 _planId) public onlyValidPlan(_planId) {
+        if (subscriptions[msg.sender].planId > 0) revert ACTIVE_SUBSCRIPTION();
+
+        Plan memory plan = plans[_planId];
+
+        if (balanceOf[msg.sender][plan.token] < plan.amount)
+            revert PAYER_IN_DEBT();
+
+        balanceOf[msg.sender][plan.token] -= plan.amount;
+        IERC20(plan.token).safeTransfer(treasury, plan.amount);
+
+        subscriptions[msg.sender] = Subscription({
+            planId: _planId,
+            expiredAt: uint48(block.timestamp + plan.period)
+        });
+
+        emit SubscriptionCreated(msg.sender, _planId);
+    }
+
+    function depositAndCreateSubscription(uint256 _planId, uint256 _amount)
+        external
+    {
+        deposit(plans[_planId].token, _amount);
+        createSubscription(_planId);
+    }
+
+    function syncSubscription(address from) public {
+        Subscription storage subscription = subscriptions[from];
+        uint256 planId = subscription.planId;
+
+        if (planId == 0) revert INACTIVE_SUBSCRIPTION();
+
+        // before expiration
+        if (block.timestamp < subscription.expiredAt) {}
+        // after expiration
+        else {
+            Plan memory plan = plans[planId];
+            uint256 count = (block.timestamp -
+                subscription.expiredAt +
+                plan.period) / plan.period;
+            uint256 amount = plan.amount * count;
+            uint256 balance = balanceOf[from][plan.token];
+
+            // not active for now
+            if (balance < amount) {
+                count = balance / plan.amount;
+                amount = plan.amount * count;
+            }
+
+            if (count > 0) {
+                balanceOf[from][plan.token] -= amount;
+                subscription.expiredAt += uint48(plan.period * count);
+
+                IERC20(plan.token).transfer(treasury, amount);
+
+                emit SubscriptionSynced(from, subscription.expiredAt);
+            }
+        }
+    }
+
+    function cancelSubscription() external {
+        syncSubscription(msg.sender);
+
+        subscriptions[msg.sender].planId = 0;
+
+        emit SubscriptionCancelled(msg.sender);
+    }
+
+    function modifySubscription(uint256 _newPlanId)
+        external
+        onlyValidPlan(_newPlanId)
+    {
+        syncSubscription(msg.sender);
```

Comment:
@bejaxer when user changes plan, we need to take into account of pro-rata calculation. Please review this article https://help.truecoach.co/en/articles/3483343-prorating-subscriptions#:~:text=Calculating%20Prorated%20Charges&text=Start%20with%20the%20total%20annual,active%20during%20that%20billing%20cycle for pro-rata calculation.